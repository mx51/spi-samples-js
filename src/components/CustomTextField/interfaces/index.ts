import { IFormEventValue } from '../../PairPage/PairForm/interfaces';

type Partial<T> = {
  [P in keyof T]?: T[P];
};

type InputProps = {
  endAdornment?: React.ReactNode;
};

type TInputProps = {
  maxLength?: number;
};

export interface ICustomTextField {
  disabled?: boolean;
  error?: boolean;
  fullWidth?: boolean;
  helperText?: string;
  id?: string;
  inputProps?: Partial<TInputProps>;
  InputProps?: Partial<InputProps>;
  label?: string;
  margin?: 'dense' | 'none';
  onBlur?: (event: IFormEventValue) => void;
  onChange?: (event: IFormEventValue) => void;
  required?: boolean;
  value?: string;
  variant?: 'filled' | 'standard' | 'outlined' | undefined;
}

export interface IErrorInputAdornment {
  isValid: boolean;
}
