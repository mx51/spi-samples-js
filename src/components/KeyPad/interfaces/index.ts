export interface IKeyPadProps {
  title: string;
  subtitle: string;
  defaultAmount: number;
  onAmountChange: (amount: number) => void;
  open: boolean;
  onClose: () => void;
}
