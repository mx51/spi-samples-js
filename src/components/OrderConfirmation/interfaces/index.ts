export interface IOrderConfirmation {
  title: string;
  pathname: string;
  editSubtotal: boolean;
}

export interface ITitleStrategy {
  Pay: string;
  Cashout: string;
  Refund: string;
}
