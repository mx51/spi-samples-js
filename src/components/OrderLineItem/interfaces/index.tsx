export interface IOrderLineItemProps {
  label: string;
  amount: number;
}
