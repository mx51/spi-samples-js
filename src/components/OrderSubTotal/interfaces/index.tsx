export interface IOrderSubTotalProps {
  label: string;
  amount: number | null;
}
