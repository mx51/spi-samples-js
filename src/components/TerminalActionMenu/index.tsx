import TerminalActionMenu from './TerminalActionMenu';

export default TerminalActionMenu;
