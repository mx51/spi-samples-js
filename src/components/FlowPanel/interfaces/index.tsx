import { ITerminalProps } from '../../../redux/reducers/TerminalSlice/interfaces';

export interface IFlowPanel {
  flow: boolean;
  receiptToggle?: boolean;
}

export interface IFlowPanelState {
  flow: boolean;
  terminal?: ITerminalProps | null;
}
