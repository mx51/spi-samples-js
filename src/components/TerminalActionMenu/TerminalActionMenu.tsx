import { IconButton, Menu, MenuItem, Typography } from '@material-ui/core';
import { Theme, createStyles, makeStyles } from '@material-ui/core/styles';
import * as React from 'react';
import { useHistory } from 'react-router-dom';
import { ReactComponent as MoreverticalIcon } from '../../images/MoreVerticalcon.svg';
import { useAppDispatch } from '../../redux/hooks';
import { ITerminalProps } from '../../redux/reducers/TerminalSlice/interfaces';
import { handleRemoveClick, handleUnPairClick } from '../../utils/common/pair/pairStatusHelpers';
import { PATH_PAIR, PATH_TERMINALS } from '../../definitions/constants/routerConfigs';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    menuPaper: {
      border: 'none',
      padding: theme.spacing(1),
      borderRadius: theme.spacing(2),
      boxShadow: `0px 16px 40px -8px ${theme.palette.secondary.main}`,
    },
    subHeading: {
      marginLeft: 16,
      fontWeight: 500,
      letterSpacing: 1,
      color: theme.palette.secondary.dark,
      marginBottom: 8,
    },
    separator: {
      borderTop: `1px solid ${theme.palette.divider}`,
    },
  })
);

const TerminalActionMenu: React.FC<{ terminal: ITerminalProps }> = ({ terminal }: { terminal: ITerminalProps }) => {
  const dispatch = useAppDispatch();

  const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null);

  const classes = useStyles();
  const open = Boolean(anchorEl);
  const history = useHistory();

  const goToTerminalDetails = (path: string) => {
    history.push(path);
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  const handleUnpair = () => {
    handleUnPairClick(dispatch, terminal.id);
    handleClose();
  };

  const handleRemove = () => {
    handleRemoveClick(dispatch, terminal.id);
    handleClose();
  };

  const handlePair = () => {
    goToTerminalDetails(`${PATH_PAIR}/${terminal.serialNumber}`);
  };
  const handleAbout = () => {
    goToTerminalDetails(`${PATH_TERMINALS}/${terminal.serialNumber}`);
  };
  return (
    <div key={`terminal-menu-${terminal.serialNumber}`}>
      <IconButton
        id={`terminal-action-button-${terminal.serialNumber}`}
        data-testid={`terminal-action-button-${terminal.serialNumber}`}
        onClick={handleClick}
        aria-label="menu"
        aria-controls={open ? 'basic-menu' : undefined}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
      >
        <MoreverticalIcon />
      </IconButton>
      <Menu
        classes={{ paper: classes.menuPaper }}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button',
        }}
      >
        <Typography variant="subtitle2" className={classes.subHeading}>
          ACTIONS
        </Typography>

        {terminal.status === 'Unpaired' && <MenuItem onClick={handlePair}>Pair terminal</MenuItem>}
        {terminal.status === 'Unpaired' && <MenuItem onClick={handleRemove}>Remove this terminal</MenuItem>}
        {['PairedConnected', 'PairedConnecting'].includes(terminal.status) && (
          <MenuItem onClick={handleUnpair}>Unpair terminal</MenuItem>
        )}
        <hr className={classes.separator} />
        <MenuItem onClick={handleAbout}>About this terminal</MenuItem>
      </Menu>
    </div>
  );
};

export default TerminalActionMenu;
