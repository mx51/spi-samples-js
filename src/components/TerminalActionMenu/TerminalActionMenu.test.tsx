import React from 'react';
import { cleanup, fireEvent, screen, waitFor } from '@testing-library/react';
import { ITerminalProps } from '../../redux/reducers/TerminalSlice/interfaces';
import { defaultAAR } from '../../definitions/constants/spiConfigs';
import { SPI_PAIR_STATUS } from '../../definitions/constants/commonConfigs';
import mockWithRedux, { defaultMockTerminals, mockTerminalInstanceId } from '../../utils/tests/common';
import TerminalActionMenu from '.';

jest.mock('../../services/spiService');

const mockTerminalConfig: ITerminalProps = {
  ...defaultMockTerminals[mockTerminalInstanceId],
  deviceAddress: defaultAAR,
  posId: 'test',
  serialNumber: mockTerminalInstanceId,
  status: SPI_PAIR_STATUS.Unpaired,
};

describe('Test <TerminalActionMenu />', () => {
  let mockContainer: Any;

  beforeEach(() => {
    mockContainer = mockWithRedux(<TerminalActionMenu terminal={mockTerminalConfig} />);
  });

  afterEach(cleanup);

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', async () => {
    expect(mockContainer).toMatchSnapshot();
  });

  test('should open action menu', async () => {
    const terminalActionButtonTestId = `terminal-action-button-${mockTerminalInstanceId}`;
    fireEvent.click(screen.getByTestId(terminalActionButtonTestId));
    const menuItemPair = await screen.getByText('Pair terminal');
    const menuItemAbout = await screen.getByText('About this terminal');

    expect(menuItemPair).toBeInTheDocument();
    expect(menuItemAbout).toBeInTheDocument();
  });
});
