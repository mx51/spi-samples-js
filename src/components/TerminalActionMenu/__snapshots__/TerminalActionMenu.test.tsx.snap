// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TerminalActionMenu /> should match snapshot 1`] = `
<div>
  <div>
    <button
      aria-haspopup="true"
      aria-label="menu"
      class="MuiButtonBase-root MuiIconButton-root"
      data-testid="terminal-action-button-123-123-123"
      id="terminal-action-button-123-123-123"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg>
          MoreVerticalcon.svg
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
